File: BusTripManagement.Models/RouteStatusResponse.cs
using System;
using System.Collections.Generic;

namespace BusTripManagement.Models
{
    public class RouteStatusResponse
    {
        public int RouteId { get; set; }
        public int NextStopIndex { get; set; }
        public List<StopStatusInfo> ReachedStops { get; set; } = new();
        public LocationInfo? LastLocation { get; set; }
    }

    public class LocationInfo
    {
        public double Latitude { get; set; }
        public double Longitude { get; set; }
        public DateTime Timestamp { get; set; }
    }

    public class StopStatusInfo
    {
        public int StopNumber { get; set; }
        public string StopName { get; set; } = string.Empty;
        public DateTime ReachedTime { get; set; }
    }
}

File: flutter_app/lib/models/route_status_response.dart
class RouteStatusResponse {
  final int routeId;
  final int nextStopIndex;
  final List<StopStatusInfo> reachedStops;
  final LocationInfo? lastLocation;

  RouteStatusResponse({
    required this.routeId,
    required this.nextStopIndex,
    required this.reachedStops,
    this.lastLocation,
  });

  factory RouteStatusResponse.fromJson(Map<String, dynamic> json) {
    return RouteStatusResponse(
      routeId: json['routeId'] as int,
      nextStopIndex: json['nextStopIndex'] as int,
      reachedStops: (json['reachedStops'] as List<dynamic>)
          .map((item) => StopStatusInfo.fromJson(item as Map<String, dynamic>))
          .toList(),
      lastLocation: json['lastLocation'] != null
          ? LocationInfo.fromJson(json['lastLocation'] as Map<String, dynamic>)
          : null,
    );
  }
}

class LocationInfo {
  final double latitude;
  final double longitude;
  final DateTime timestamp;

  LocationInfo({
    required this.latitude,
    required this.longitude,
    required this.timestamp,
  });

  factory LocationInfo.fromJson(Map<String, dynamic> json) {
    return LocationInfo(
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      timestamp: DateTime.parse(json['timestamp'] as String),
    );
  }
}

class StopStatusInfo {
  final int stopNumber;
  final String stopName;
  final DateTime reachedTime;

  StopStatusInfo({
    required this.stopNumber,
    required this.stopName,
    required this.reachedTime,
  });

  factory StopStatusInfo.fromJson(Map<String, dynamic> json) {
    return StopStatusInfo(
      stopNumber: json['stopNumber'] as int,
      stopName: json['stopName'] as String,
      reachedTime: DateTime.parse(json['reachedTime'] as String),
    );
  }
}
